lib/client/keystore.go

```diff
@@ -826,79 +685,129 @@ func (noLocalKeyStore) AddKnownHostKeys(hostname string, keys []ssh.PublicKey) e
+	return nil, errNoLocalKeyStore
+}
 
 // MemLocalKeyStore is an in-memory session keystore implementation.
 type MemLocalKeyStore struct {
 	fsLocalNonSessionKeyStore
-	inMem map[keyIndex]*Key
+	inMem memLocalKeyStoreMap
 }
 
-// NewMemLocalKeyStore initializes a MemLocalKeyStore, the key directory here is only used
-// for storing CA certificates and known host fingerprints.
+// memLocalKeyStoreMap is a three-dimensional map indexed by [proxyHost][username][clusterName]
+type memLocalKeyStoreMap = map[string]map[string]map[string]*Key
+
+// NewMemLocalKeyStore initializes a MemLocalKeyStore.
+// The key directory here is only used for storing CA certificates and known
+// host fingerprints.
 func NewMemLocalKeyStore(dirPath string) (*MemLocalKeyStore, error) {
 	dirPath, err := initKeysDir(dirPath)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
-
-	inMem := make(map[keyIndex]*Key)
-	return &MemLocalKeyStore{fsLocalNonSessionKeyStore: fsLocalNonSessionKeyStore{
-		log: logrus.WithFields(logrus.Fields{
-			trace.Component: teleport.ComponentKeyStore,
-		}),
-		KeyDir: dirPath,
-	}, inMem: inMem}, nil
+	return &MemLocalKeyStore{
+		fsLocalNonSessionKeyStore{
+			log:    logrus.WithField(trace.Component, teleport.ComponentKeyStore),
+			KeyDir: dirPath,
+		},
+		memLocalKeyStoreMap{},
+	}, nil
 }
 
 // AddKey writes a key to the underlying key store.
-func (s *MemLocalKeyStore) AddKey(proxy string, username string, key *Key) error {
-	s.inMem[keyIndex{proxyHost: proxy, username: username}] = key
-	if key.ClusterName != "" {
-		s.inMem[keyIndex{proxyHost: proxy, username: username, clusterName: key.ClusterName}] = key
+func (s *MemLocalKeyStore) AddKey(key *Key) error {
+	if err := key.KeyIndex.CheckAllSet(); err != nil {
+		return trace.Wrap(err)
+	}
+	_, ok := s.inMem[key.ProxyHost]
+	if !ok {
+		s.inMem[key.ProxyHost] = map[string]map[string]*Key{}
 	}
+	_, ok = s.inMem[key.ProxyHost][key.Username]
+	if !ok {
+		s.inMem[key.ProxyHost][key.Username] = map[string]*Key{}
+	}
+	s.inMem[key.ProxyHost][key.Username][key.ClusterName] = key
 	return nil
 }
 
-// GetKey returns the session key for the given username and proxy.
-func (s *MemLocalKeyStore) GetKey(proxy, username string, opts ...KeyOption) (*Key, error) {
-	idx := keyIndex{proxyHost: proxy, username: username}
-	entry, ok := s.inMem[idx]
-	if !ok {
-		return nil, trace.NotFound("key for %v not found", idx)
+// GetKey returns the user's key including the specified certs.
+func (s *MemLocalKeyStore) GetKey(idx KeyIndex, opts ...CertOption) (*Key, error) {
+	var key *Key
+	if idx.ClusterName == "" {
+		// If clusterName is not specified then the cluster-dependent fields
+		// are not considered relevant and we may simply return any key
+		// associated with any cluster name whatsoever.
+		for _, found := range s.inMem[idx.ProxyHost][idx.Username] {
+			key = found
+			break
+		}
+	} else {
+		key = s.inMem[idx.ProxyHost][idx.Username][idx.ClusterName]
```

Comment:
Check nested maps aren't nil to avoid panics.

Reply:
An uninitialized or `nil` map returns the zero value for every key. It can't panic here.